File: benchmarks/binarytrees/binarytrees.zig
const std = @import("std");

const Node = struct {
    l: ?*Node,
    r: ?*Node,
};

fn makeTree(alloc: std.mem.Allocator, d: i32) !*Node {
    const n = try alloc.create(Node);
    if (d <= 0) {
        n.l = null;
        n.r = null;
        return n;
    }
    n.l = try makeTree(alloc, d - 1);
    n.r = try makeTree(alloc, d - 1);
    return n;
}

fn check(n: *Node) i32 {
    if (n.l == null) return 1;
    return 1 + check(n.l.?) + check(n.r.?);
}

fn freeTree(alloc: std.mem.Allocator, n: *Node) void {
    if (n.l) |l| freeTree(alloc, l);
    if (n.r) |r| freeTree(alloc, r);
    alloc.destroy(n);
}

pub fn main() !void {
    const alloc = std.heap.smp_allocator;
    const depth: i32 = 15;
    const t = try makeTree(alloc, depth);
    const c = check(t);
    freeTree(alloc, t);
    std.debug.print("depth {d} check={d}\n", .{ depth, c });
}

File: benchmarks/binarytrees/binarytrees.rs
enum Tree {
    Leaf,
    Node(Box<Tree>, Box<Tree>),
}

fn make_tree(d: i32) -> Box<Tree> {
    if d <= 0 {
        return Box::new(Tree::Leaf);
    }
    Box::new(Tree::Node(make_tree(d - 1), make_tree(d - 1)))
}

fn check(t: &Tree) -> i32 {
    match t {
        Tree::Leaf => 1,
        Tree::Node(l, r) => 1 + check(l) + check(r),
    }
}

fn main() {
    let depth: i32 = 15;
    let t = make_tree(depth);
    let c = check(&t);
    println!("depth {} check={}", depth, c);
}
